File: 0300-Longest_Increasing_Subsequence/Longest_Increasing_Subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] counts = new int[nums.length];
        int max = 1;
        counts[nums.length - 1]++;
        
        for(int i = nums.length - 2; i >= 0; i--){
            int count = 1;
            for(int j = nums.length - 1; j > i; j--){
                if(nums[i] < nums[j]){
                    count = Math.max(count, (counts[j] + 1));
                }
            }
            counts[i] = count;
            max = Math.max(count, max);
        }
        
        return max;
    }
}

File: 0300-LongestIncreasingSubsequence/LongestIncreasingSubsequence.go
package longestincreasingsubsequence

func lengthOfLIS(nums []int) int {
	dp := make([]int, len(nums))
	dp[len(nums)-1] = 1

	ans := 1
	for i := len(nums) - 2; i >= 0; i-- {
		count := 1
		for j := len(nums) - 1; j > i; j-- {
			if nums[i] < nums[j] {
				count = max(count, dp[j]+1)
			}
		}
		dp[i] = count
		ans = max(ans, count)
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
